packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -224,6 +227,22 @@ export function evaluateTemplate(vm: VM, html: Template): VNodes {
     return vnodes;
 }
 
+function validateTemplateVersion(template: Template, vm: VM) {
+    // Validate that the template was compiled with the same version of the LWC compiler used for the runtime engine.
+    // Note this only works in unminified dev mode because it relies on code comments.
+    if (template === defaultEmptyTemplate) {
+        // skip the default template
+        return;
+    }
+    checkVersionMismatch(template, vm);
+    if (!isUndefined(template.stylesheets)) {
+        for (const stylesheet of flattenStylesheets(template.stylesheets)) {
+            // Verify that the stylesheet was compiled with a compatible LWC version
+            checkVersionMismatch(stylesheet, vm);
```

Comment:
This could also be moved to `registerTemplate` to surface the error during module evaluation.

Reply:
The problem with using `registerComponent` and `registerTemplate` is that we lose the `VM`, which is helpful information to put in the warning. But I agree it's still better to throw the error as early as possible, so let's move the logic there.